### DISCUSSION TITLE
LEGO 42146 Liebherr LR 1300 crane with one remote connected to each Technic hub

### DISCUSSION THREAD

**Discussion by BertLindeman** [pybricks/support#1259] [Show and tell]
### General
Have built the huge `LEGO 42146 Liebherr LR 1300 crane` and "played" with a mobile phone on the Powered up app.
Great build that took not enough time 😃 

### Rationale
I want to use pybricks.
Have used `beta.pybricks.com` for this project.

The crane is controlled by two Technic hubs.
In pybricks the remote `LEGO 28739 Control Unit, Powered Up` cannot be connected to more than one hub.

So: one remote controls one Technic Hub.
The crane has two hubs, so I need two remotes. (Although hub to hub communication is another possibility)

### Program description

- lego_42146_set_remote_names.py 
  Utility program to set the name of the remote 
  Run two times. Once for each remote.

- lego_42146_liebherr_config.py
  Module that keeps various values for the main program, like the name of the top and bottom hub, names of the remotes.

- lego_42146_main.py
  The program to run at least once on each hub, to load the specific module for that hub.
  Module loaded on the top hub: lego_42146_tophub.py
  Module loaded on the bottom hub: lego_42146_bottomhub.py
  The hubs remember the program that ran last on it.

#### Prepare 
- Download the zip file from [Bricksafe lego_42146_pybricks.zip](https://bricksafe.com/files/Bert_bricksafe/42146-lr-13000/lego_42146_pybricks.zip)
- unzip the zip
- read the README.txt
- import all lego_42146*.py programs into a session on beta.pybricks.com
- adapt if you want `lego_42146_liebherr_config.py`
- run `lego_42146_set_remote_names.py` while connected to any Technic hub to set the remote names (so run twice).
- Run the main module `lego_42146_main.py` once on each hub.
   After the program was orderly stopped, the crane can be used without the use of a PC.

#### Start both hubs (one at a time)
- short press the green button on the `Bottom hub` (starts blinking)
- short press the green button on the `Right remote` (starts blinking)
- short press the green button on the `Bottom hub`
  The hub goes to steady blue and then after connecting to the remote they both turn on brown.

One hub active, now the other:
- short press the green button on the `Top hub` (starts blinking)
- short press the green button on the `Left remote` (starts blinking)
- short press the green button on the `Top hub`
  The hub goes to steady blue and then after connecting to the remote they both turn on green.

**Ready to play**

### The remote units
![image](https://github.com/pybricks/support/assets/8142081/409a304d-22c5-437e-ad91-72f383fd1eb8)

#### Functions of the remotes:
-    Both remotes:

        CENTER (green) button: Send a STOP command to the connected hub; the program stops

-    Left remote   Top Hub   State `green`:

|Key|action|
|----|----|
|Left PLUS: |Hoist UP|
|Left CENTER (red):|Switch lights (ON / OFF)|
|Left MINUS:         |Hoist DOWN|
|  |  |
|Right PLUS:           |Boom OUT|
|Right CENTER (red):  |Switch color mode of Left remote (Green to Yellow)|
|Right MINUS:      |Boom IN|

-    Left remote   Top Hub   State `yellow`:

|Key|action|
|---|---|
|Left PLUS:            |Hoist UP|
|        Left CENTER (red):    |Switch lights (ON / OFF)|
|        Left MINUS:           |Hoist DOWN|
| | |
| Right PLUS:           |Jib OUT|
| Right CENTER (red):   |Switch color mode of Left remote (Yellow to Green)|
| Right MINUS:          |Jib IN|

-    Right remote  Bottom Hub  State `brown`:

|Key|action|
|----|----|
|Left PLUS:            |Drive straight Forward (Direction where the hub is)|
|        Left CENTER (red):    |.|
|        Left MINUS:           |Drive straight Backward|
|||
|        Right PLUS:           |Drive and turn to the left|
|        Right CENTER (red):   |Switch color mode of Right remote (Brown to Magenta)|
|        Right MINUS:          |Drive and turn to the right|

-    Right remote  Bottom Hub  State `magenta`:

|Key|action|
|----|----|
|        Left PLUS:            |Rotate Superstructure to the Left}
|        Left CENTER (red):    |.|
|        Left MINUS:           |Rotate Superstructure to the Right|
|||
|        Right PLUS:           |.|
|        Right CENTER (red):   |Switch color mode of Right remote (Magenta to Brown)|
|        Right MINUS:          |.|

The hoist and rotate motors have been set to start REALLY SLOW as in the real world.
So have a bit of patience with these functions.
If you keep the button pressed, the hoist speed will amaze you in the end. (I think.)

I feel that still information is missing here, but at the moment I do not know what.
The information is also in the README.txt file, possibly in other words.

Please, let me know if you miss something, or spot errors.

Thank you,

Bert

**Comment by laurensvalk** [comment]
Nice! I was playing with that just yesterday :smile: Looking forward to seeing your results.

![image](https://github.com/pybricks/support/assets/12326241/49dcdee3-e320-4554-8989-4b694ead3edc)

**Reply by BertLindeman** [reply]
As am I in yours! 😃

**Comment by BertLindeman** [comment]
The comments in `lego_42146_main.py` state hub names that are different from the names in `lego_42146_liebherr_config.py`
**The names in the configuration module are leading in this.**

<details>

<summary>The main program: &emsp;lego_42146_main.py.<br /> 
&emsp;Click to show the source</summary>

```python

"""
Have split the program because the left hub does not use the code for the right hub and vice versa.
Maybe it helps to ease a bit on memory usage and time needed to load.

Run lego_42146_main.py for both hubs.
For each hub:
    (if the hub is OFF, press the green button until the led blinks blue)
    press the green center button on the correct remote for this hub
    short press the green button on the hub, so the led shows a in and out fading blue led
    The color of the remote and the hub will change to a matching color to indicate "connected".


This main program imports (loads and runs) either
    lego_42146_tophub.py
or  lego_42146_bottomhub.py
depending on the paired hubname

In this case the names of the hubs are supposed to be:
    "techHubC"  for the bottom hub
    "techHubD"  for the top hub

    The hub name can be set during flashing of the pybricks firmware to the hub.

    This set of programs is tested with
    version  ('technichub', '3.3.0b6', 'v1.20.0-23-g6c633a8dd on 2023-06-02')

"""

# from pybricks.hubs import ThisHub
from pybricks.hubs import TechnicHub
from pybricks.parameters import Color
from pybricks.tools import wait

from pybricks import version

import lego_42146_liebherr_config

HUB = TechnicHub()
# pylint complains probably due to a pybrickdev and beta.pybricks.com difference
hubname = HUB.system.name()
print("\tThis HUB name is '" + hubname + "' at version ", version)

HUBNAMES = lego_42146_liebherr_config.HUBNAMES

TOPHUB = HUBNAMES["top"]
BOTTOMHUB = HUBNAMES["bottom"]

if hubname == TOPHUB:
    HUB = None  # release the hub and define again in the separately imported program.
    EXPECT_REMOTE_NAME = "Handset left"
    # print("import lego_42146_tophub to connect to remote", EXPECT_REMOTE_NAME)
    import lego_42146_tophub  # noqa pylint: disable=W0611
elif hubname == BOTTOMHUB:
    HUB = None  # release the hub and define again in the separately imported program.
    EXPECT_REMOTE_NAME = "Handset right"
    # print("import lego_42146_tophub to connect to remote", EXPECT_REMOTE_NAME)

    import lego_42146_bottomhub  # noqa pylint: disable=W0611
else:
    # from pybricks.hubs import TechnicHub
    # HUB = TechnicHub()
    HUB.light.animate([Color.RED, Color.NONE], interval=128)
    print("\n\tThis hub '" + hubname + "' is not an expected HUB", end="")
    print("\n\tlike top hub    '" + TOPHUB + "'", end="")
    print("\n\tor   bottom hub '" + BOTTOMHUB + "'\n")
print("program Done\nWaiting 10 seconds to show the error")
wait(10000)

```

</details>

<details>

<summary>The configuration module &emsp;lego_42146_liebherr_config.py*.<br /> 
&emsp;Click to show the source</summary>

```python
# liebherr config module

"""
Tested with beta.pybricks.com and firmware version 3.3.0b8

Unzip the lego_42146.zip in a separate directory, if a zip is downloaded from bricksafe.
Load all lego_42146_*.py programs into a beta.pybricks.com session (into e. g. Google Chrome).

Connections and hub names:

  Top hub     devices connected to remote left:
    techHubD
      Port.A  Technic Large Angular Motor
      Port.B  Technic Large Angular Motor
      Port.C                Power Leds (Optional)
      Port.D  Technic Large Angular Motor

  Bottom hub  devices connected to remote right:
    techHubC
      Port.A  ---
      Port.B  Technic Large Angular Motor
      Port.C  Technic Large Angular Motor
      Port.D  Technic Large Angular Motor



  Just for fun: Names of the main elements of the crane:

  Crawler Tracks:
    Crawler cranes are mounted on a set of crawler tracks or caterpillar tracks
    that allow the crane to move over various types of terrain.

  Upper Carriage or Superstructure:
    The upper carriage houses the crane's engine, control cabin, and other components.
    It is responsible for the crane's rotation and lifting operations.

  Boom:
    The boom is the main structure of the crane that extends upward and outward.
    It is used for lifting and lowering heavy loads.

  Jib:
    A jib is an additional arm or extension that can be attached to the boom to increase the crane's reach and lifting capacity.
    It can be luffed (raised or lowered) to adjust the angle and reach of the load.

  Counterweights:
    Large crawler cranes often require counterweights to balance the crane when lifting heavy loads.
    These counterweights are typically attached to the rear of the crane.

  Hook and Block:
    The hook and block assembly is used to attach and hoist loads.
    The hook is connected to the load,
    and the block contains multiple sheaves or pulleys to provide the lifting power.

  Hoist Winch:
    The hoist winch is a powerful winching system responsible for raising and lowering the hook and load.

  Boom Hoist Winch:
    The boom hoist winch is used to control the elevation or angle of the boom.

"""
# enter the names you gave the hubs at the moment of firmware install
# enter the names of the hand sets
# Both Hub and Remote names are CASE-sensitive!
HUBNAMES = {"top": "42146top", "bottom": "42146bottom"}  #  <<< update if you want to; Names MUST be different.

# You can set the names of the two remotes by running lego_42146_set_remote_names.py
REMOTENAMES = {"left": "Handset left", "right": "Handset right"}  #  <<< update if you want to; Names MUST be different.

# Speeds and accelerations to use
STRAIGHT_SPEED = 300     # straight driving speed          # 300 was a bit too much maybe OK with slow acceleration
TURN_RATE = 600          # turning rate                    # 300 is good with a very slow acceleration
ROTATESPEED = 1200       # motor speed to turn the superstructure
BOOMSPEED = 400          # motor speed of the winch that controls the movement of the main boom.
JIBSPEED = 400           # motor speed of the winch that controls the luffing (raising or lowering) of the jib.
HOISTSPEED = 1500        # motor speed of the winch that controls hoisting.

MY_SUPERSTR_ACCEL = 200  # a very low acceleration so it starts to turn slowly
MY_DRIVE_ACCEL = 50      # a very low acceleration so it starts to drive a bit slowly
MY_HOIST_ACCEL = 50      # a bit slower start of hoisting

# status colors right remote:
STATUSBROWN = 0          # state brown   - default is brown for the right remote
STATUSMAGENTA = 1        # state magenta
# status colors  left remote:
STATUSGREEN = 2          # state green   - default is green for the left remote
STATUSYELLOW = 3         # state yellow

STATE_COLOR = {0: "Brown",
               1: "Magenta",
               2: "Green",
               3: "Yellow",
               }

```

</details>
